# Exploratory Data Analysis

**Dataset**

We gather the data for the dataset by scraping Twitter's tweet that contain words about incidents of sexual violence and harassment from the victim. We scrap #MeToo hashtag which is a social movement against sexual abuse and sexual harassment where people publicize allegations of sex crimes. The phrase "Me Too" was initially used in this context on social media in 2006, on Myspace, by sexual harassment survivor and activist Tarana Burke.

We collected the data with 4 period time.


1.   1st period: 15 April - 23 April 2021
2.   2nd period: 23 April - 29 April 2021
3.   3rd period: 29 April - 05 May 2021
4.   4th period: 05 May - 11 May 2021



In [40]:
#Imports
import pandas as pd
import numpy as np
import re

# 1st period (15 Apr - 23 Apr 2021)

In [41]:
#Import First Dataset
url_one = 'https://drive.google.com/file/d/1lTvQke_pO35JgAj1Bzj5ULsA_gno94DQ/view?usp=sharing'
file_id_one = url_one.split('/')[-2]
csv_url_one = 'https://drive.google.com/uc?id=' + file_id_one
df_one = pd.read_csv(csv_url_one)
df_one.head()

index  ...                                               text
0      1  ...  b\'"The ironies of the #MeToo movement are too...
1      2  ...  RT @WomenReadWomen: Women in Kuwait have launc...
2      3  ...  RT @j_andreamolina: \\xe2\\x9a\\xa0\\xef\\xb8\...
3      4  ...  RT @SammiStitches: Charges brought against the...
4      5  ...  b"And the fourth presenter on 26th May is Hann...

[5 rows x 4 columns]

In [42]:
print("Shape of 1st period data-> ", df_one.shape)

Shape of 1st period data->  (6600, 4)


# 2nd period (23 Apr - 29 Apr 2021)

In [43]:
#Import Second Dataset
url_two = 'https://drive.google.com/file/d/1xMEtidbapgBLs82ijoJvk0FsoPlL76yG/view?usp=sharing'
file_id_two = url_two.split('/')[-2]
csv_url_two = 'https://drive.google.com/uc?id=' + file_id_two
df_two = pd.read_csv(csv_url_two)
df_two.head()

index  ...                                               text
0      1  ...  b\'RT @ThatUmbrella: Advocates of #metoo shoul...
1      2  ...  Boo another report while casualties pile up! Y...
2      3  ...  b"RT @MsAmyMacPherson: 1/9 There\'s a SUPER HU...
3      4  ...  b"RT @RaquelDancho: Important to note: if the ...
4      5  ...  RT @Aspasia_1: Basically .@MetOpera could have...

[5 rows x 4 columns]

In [44]:
print("Shape of 2nd period data-> ", df_two.shape)

Shape of 2nd period data->  (8900, 4)


# 3rd period (29 Apr - 05 May 2021)

In [45]:
#Import Third Dataset
url_three = 'https://drive.google.com/file/d/1vNVBMchza3792vB16pGQwzHf65eJYlOB/view?usp=sharing'
file_id_three = url_three.split('/')[-2]
csv_url_three = 'https://drive.google.com/uc?id=' + file_id_three
df_three = pd.read_csv(csv_url_three)
df_three.head()

index  ...                                               text
0      1  ...  RT @BombshellDAILY: REPUBLICAN BRUTALLY RAPES ...
1      2  ...  b"Anthony Kiedis mocked those fighting Bin Lad...
2      3  ...  RT @edrormba: Palestinian-American brings #MeT...
3      4  ...  RT @bette_oh: And, just like that, the #MeToo ...
4      5  ...  This guy was in #MeToo right?? https://t.co/zZ...

[5 rows x 4 columns]

In [46]:
print("Shape of 3rd period data-> ", df_three.shape)

Shape of 3rd period data->  (7100, 4)


# 4th period (05 May - 11 May 2021)

In [ ]:
#Import Fourth Dataset
url_three = ''
file_id_four = url_four.split('/')[-2]
csv_url_four = 'https://drive.google.com/uc?id=' + file_id_four
df_four = pd.read_csv(csv_url_four)
df_four.head()
#SOON

In [ ]:
print("Shape of 4th period data-> ", df_four.shape)

# Combine All Dataset

After we collected all data within time period, we combined it into a new dataframe. There are 4 columns: 

1.   'index' : index of tweets
2.   'user_id' : user ID that make the tweet
3.   'timestamp' : time when the user tweet 
4.   'text' : tweets



In [48]:
#df = pd.concat([df_one, df_two, df_three, df_four])
df = pd.concat([df_one, df_two, df_three])

In [49]:
df.head()

index  ...                                               text
0      1  ...  b\'"The ironies of the #MeToo movement are too...
1      2  ...  RT @WomenReadWomen: Women in Kuwait have launc...
2      3  ...  RT @j_andreamolina: \\xe2\\x9a\\xa0\\xef\\xb8\...
3      4  ...  RT @SammiStitches: Charges brought against the...
4      5  ...  b"And the fourth presenter on 26th May is Hann...

[5 rows x 4 columns]

In [50]:
print("Combined shape of all data-> ", df.shape)

Combined shape of all data->  (22600, 4)


In [51]:
# Check null
df.isnull().sum()

index        0
user_id      0
timestamp    0
text         0
dtype: int64

In [52]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 22600 entries, 0 to 7099
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   index      22600 non-null  int64 
 1   user_id    22600 non-null  object
 2   timestamp  22600 non-null  object
 3   text       22600 non-null  object
dtypes: int64(1), object(3)
memory usage: 882.8+ KB


# Drop Unnecessary Columns

We will use 'text' column to get the tweet because our analysis and project will be based on this column.

In [53]:
df = df.drop(columns=['index', 'user_id', 'timestamp'])
df

text
0     b\'"The ironies of the #MeToo movement are too...
1     RT @WomenReadWomen: Women in Kuwait have launc...
2     RT @j_andreamolina: \\xe2\\x9a\\xa0\\xef\\xb8\...
3     RT @SammiStitches: Charges brought against the...
4     b"And the fourth presenter on 26th May is Hann...
...                                                 ...
7095  RT @brianlilley: In QP today, \\xe2\\x81\\xa6@...
7096  RT @shades_dirty: Stop police sexual violence ...
7097  RT @GretchenCarlson: I\\xe2\\x80\\x99ll speak ...
7098  RT @brianlilley: In QP today, \\xe2\\x81\\xa6@...
7099  RT @SenatorHousakos: In fairness, he also didn...

[22600 rows x 1 columns]

# Data Cleaning

Here we can see for some tweets that there are still many contractions like 'RT', user ID, URL, and punctuations.

In [54]:
for index,text in enumerate(df['text'][50:60]):
  print('Tweets %d:\n'%(index+1),text)

Tweets 1:
 RT @SammiStitches: Charges brought against the guy that smacked my ass: sexual assault and battery. Fuck around and find out. \\xf0\\x9f\\x91\\x8a#sexualAssau\\xe2\\x80\\xa6'
Tweets 2:
 RT @SammiStitches: Charges brought against the guy that smacked my ass: sexual assault and battery. Fuck around and find out. \\xf0\\x9f\\x91\\x8a#sexualAssau\\xe2\\x80\\xa6'
Tweets 3:
 RT @KellyPlu: Harassment, retaliation &amp
Tweets 4:
 Looking forward 2 a panel chat @CityCynthia today about systemic violence of police women in Canada. I am grateful the team of women on the panel &amp
Tweets 5:
 @ArunKrishnan_ This b...d is in #MeToo case, does not talk about it.'
Tweets 6:
 This is the child that was born as a result of a statutory rape committed by NLRPD cop Tommy Norman. The mother was 15 when Officer Norman had sex with her and got her pregnant. NLRPD needs to address
Tweets 7:
 RT @GIWLkings: EVENT: A new cultural reckoning? Gendered violence &amp
Tweets 8:
 RT @gecko39: \\xe2\\x80\\x

In [55]:
#Safe Clean (Cleaning the new line command before it get removed by the next regex)
df['text'] =  [re.sub(r"\\n", " ", str(j)) for j in df['text']]

#Cleaning the 'b', Retweet, Mentions, Unicode (\x08\,\xef\,etc), and punctuation marks.
df['text'] =  [re.sub(r"^b|http\S+|[0-9]\/[0-9]|\'|\"|\:|\;|\.|\,|RT|\\\\x[\w][\w]|\@[\w\_]*|\#[\w]*|\\", 
                           "", str(j)) for j in df['text']]

#Cleaning the blank space at the start of the sentences after the 1st cleaning process, hashtag and other punctuation that remains
df['text'] =  [re.sub(r"^\s*|\=|\-|\#[\w]*\s?|[a-z]*\#[\w]*\s?|[\(\)\|\!\$\?\*]*|\&amp", "", str(j)) for j in df['text']]

#Cleaning digits and words containing digits
#df['text'] = [re.sub('\w*\d\w*','', str(j)) for j in df['text']]

#Adding space after slash
df['text'] =  [re.sub(r"\/", " ", str(j)) for j in df['text']]
df.head()

text
0  The ironies of the  movement are too painfully...
1  Women in Kuwait have launched a social media c...
2  Triger Warning Raising Awareness of Sexual Ass...
3  Charges brought against the guy that smacked m...
4  And the fourth presenter on 26th May is Hannah...

In [56]:
#Transform the text into lowercase
df['text'] = df['text'].str.lower()

In [57]:
#Drop any duplicate row
df = df.drop_duplicates()

In [58]:
#Fill up the empty cells
df['text'].replace('', np.nan, inplace=True)

/usr/local/lib/python3.7/dist-packages/pandas/core/series.py:4582: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  method=method,


As we can see now, all the constractions already removed and the text already transformed into lowercase text.

In [59]:
for index,text in enumerate(df['text'][50:60]):
  print('Tweets %d:\n'%(index+1),text)

Tweets 1:
 he also had this app on his phones its used to hide messages within jpg ima
Tweets 2:
 just one question where is   english teacher marka bodine 31 sexually abused boy under 14 for three years unt
Tweets 3:
 these ladies have been facing sexual harassment and brutally tortured during interrogation    
Tweets 4:
 where tf is    high school teacher monica young 24 admitted to repeatedly sexually assaulting a 14yearold hi
Tweets 5:
 starship  sara 1985 grunt records on youtube      
Tweets 6:
 the  movement including the brave women who have gone to court and the debut of the civil code have reinvigora
Tweets 7:
 he also had this app on his phones its used to hide messages within jpg images     
Tweets 8:
 the  movement including the brave women who have gone to court and the debut of the civil code have reinvigorated discussions about sexual harassment in china   
Tweets 9:
 actually amazing to see  standing up to be counted with the  and decolonisation sessions massive respec

In [60]:
#Create Index
df['index'] = range(0, len(df.index), 1)
df.head()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


text  index
0  the ironies of the  movement are too painfully...      0
1  women in kuwait have launched a social media c...      1
2  triger warning raising awareness of sexual ass...      2
3  charges brought against the guy that smacked m...      3
4  and the fourth presenter on 26th may is hannah...      4

In [62]:
#creating the word count column
df['word_count'] = df['text'].apply(lambda x: len(str(x).split()))
df.head()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


text  index  word_count
0  the ironies of the  movement are too painfully...      0          31
1  women in kuwait have launched a social media c...      1          18
2  triger warning raising awareness of sexual ass...      2          11
3  charges brought against the guy that smacked m...      3          18
4  and the fourth presenter on 26th may is hannah...      4          32

In [36]:
# Shape of cleaned data
print("Shape of cleaned data-> ", df.shape)

Shape of cleaned data->  (8738, 2)


Shape of each period data:


1.   1st period : 6600
2.   2nd period : 8900
3.   3rd period : 7100
4.   4th period : 

**TOTAL: 22600 tweets**

After we cleaned the data, we only get **8738** tweets that we can use for training model.



# Stopwords

(ini pake Spacy stopwords) Versi 1: https://www.analyticsvidhya.com/blog/2020/04/beginners-guide-exploratory-data-analysis-text-data/

In [71]:
# Dictionary of English Contractions
contractions_dict = { "ain't": "are not","'s":" is","aren't": "are not",
                     "can't": "cannot","can't've": "cannot have",
                     "'cause": "because","could've": "could have","couldn't": "could not",
                     "couldn't've": "could not have", "didn't": "did not","doesn't": "does not",
                     "don't": "do not","hadn't": "had not","hadn't've": "had not have",
                     "hasn't": "has not","haven't": "have not","he'd": "he would",
                     "he'd've": "he would have","he'll": "he will", "he'll've": "he will have",
                     "how'd": "how did","how'd'y": "how do you","how'll": "how will",
                     "I'd": "I would", "I'd've": "I would have","I'll": "I will",
                     "I'll've": "I will have","I'm": "I am","I've": "I have", "isn't": "is not",
                     "it'd": "it would","it'd've": "it would have","it'll": "it will",
                     "it'll've": "it will have", "let's": "let us","ma'am": "madam",
                     "mayn't": "may not","might've": "might have","mightn't": "might not", 
                     "mightn't've": "might not have","must've": "must have","mustn't": "must not",
                     "mustn't've": "must not have", "needn't": "need not",
                     "needn't've": "need not have","o'clock": "of the clock","oughtn't": "ought not",
                     "oughtn't've": "ought not have","shan't": "shall not","sha'n't": "shall not",
                     "shan't've": "shall not have","she'd": "she would","she'd've": "she would have",
                     "she'll": "she will", "she'll've": "she will have","should've": "should have",
                     "shouldn't": "should not", "shouldn't've": "should not have","so've": "so have",
                     "that'd": "that would","that'd've": "that would have", "there'd": "there would",
                     "there'd've": "there would have", "they'd": "they would",
                     "they'd've": "they would have","they'll": "they will",
                     "they'll've": "they will have", "they're": "they are","they've": "they have",
                     "to've": "to have","wasn't": "was not","we'd": "we would",
                     "we'd've": "we would have","we'll": "we will","we'll've": "we will have",
                     "we're": "we are","we've": "we have", "weren't": "were not","what'll": "what will",
                     "what'll've": "what will have","what're": "what are", "what've": "what have",
                     "when've": "when have","where'd": "where did", "where've": "where have",
                     "who'll": "who will","who'll've": "who will have","who've": "who have",
                     "why've": "why have","will've": "will have","won't": "will not",
                     "won't've": "will not have", "would've": "would have","wouldn't": "would not",
                     "wouldn't've": "would not have","y'all": "you all", "y'all'd": "you all would",
                     "y'all'd've": "you all would have","y'all're": "you all are",
                     "y'all've": "you all have", "you'd": "you would","you'd've": "you would have",
                     "you'll": "you will","you'll've": "you will have", "you're": "you are",
                     "you've": "you have"}

# Regular expression for finding contractions
contractions_re=re.compile('(%s)' % '|'.join(contractions_dict.keys()))

# Function for expanding contractions
def expand_contractions(text,contractions_dict=contractions_dict):
  def replace(match):
    return contractions_dict[match.group(0)]
  return contractions_re.sub(replace, text)

# Expanding Contractions in the reviews
df['text']=df['text'].apply(lambda x:expand_contractions(str(x)))


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:50: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


In [72]:
# Importing spacy
import spacy

# Loading model
nlp = spacy.load('en_core_web_sm',disable=['parser', 'ner'])

# Lemmatization with stopwords removal
df['lemmatized']=df['text'].apply(lambda x: ' '.join([token.lemma_ for token in list(nlp(x)) if (token.is_stop==False)]))

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [75]:
df_grouped=df[['index','lemmatized']].groupby(by='index').agg(lambda x:' '.join(x))
df_grouped.head()

lemmatized
index                                                   
0      irony   movement painfully obvious spearhead h...
1      woman kuwait launch social medium campaign dem...
2      triger warning raise awareness sexual assault ...
3      charge bring guy smack ass sexual assault batt...
4      fourth presenter 26th hannah baylor phd studen...

In [77]:
# Creating Document Term Matrix
from sklearn.feature_extraction.text import CountVectorizer
cv=CountVectorizer(analyzer='word')
data=cv.fit_transform(df_grouped['lemmatized'])
df_dtm = pd.DataFrame(data.toarray(), columns=cv.get_feature_names())
df_dtm.index=df_grouped.index
df_dtm.head(3)

000  0021  01  03  0551  06  ...  zoe  zoepost  zone  zoom  zu  zurich
index                               ...                                      
0        0     0   0   0     0   0  ...    0        0     0     0   0       0
1        0     0   0   0     0   0  ...    0        0     0     0   0       0
2        0     0   0   0     0   0  ...    0        0     0     0   0       0

[3 rows x 12096 columns]

In [78]:
# Importing wordcloud for plotting word clouds and textwrap for wrapping longer text
from wordcloud import WordCloud
from textwrap import wrap

# Function for generating word clouds
def generate_wordcloud(data,title):
  wc = WordCloud(width=400, height=330, max_words=150,colormap="Dark2").generate_from_frequencies(data)
  plt.figure(figsize=(10,8))
  plt.imshow(wc, interpolation='bilinear')
  plt.axis("off")
  plt.title('\n'.join(wrap(title,60)),fontsize=13)
  plt.show()
  
# Transposing document term matrix
df_dtm=df_dtm.transpose()

# Plotting word cloud for each product
for index,product in enumerate(df_dtm.columns):
  generate_wordcloud(df_dtm[product].sort_values(ascending=False),product)

ZeroDivisionError: ignored

(ini pake nltk stopwords)
Versi 2: https://neptune.ai/blog/exploratory-data-analysis-natural-language-processing-tools

Versi 3: https://github.com/briannalytle/reddit_nlp

# Plot

# Conclusion